src/main/java/com/example/onlinebookstore/service/shopping/cart/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,101 @@
+import jakarta.transaction.Transactional;
+import java.util.HashSet;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final UserRepository userRepository;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto findById(Long id) {
+        return shoppingCartMapper.toDto(shoppingCartRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException(
+                        "Cannot find shoppingCart by id " + id)
+                )
+        );
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addToCart(Long id, CreateCartItemRequestDto requestDto) {
+        User user = userRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Cannot find user")
+        );
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(id)
+                .orElseGet(() -> {
+                    ShoppingCart shoppingCartNew = new ShoppingCart();
+                    shoppingCartNew.setUser(user);
+                    shoppingCartNew.setCartItems(new HashSet<>());
+                    return shoppingCartRepository.save(shoppingCartNew);
+                }
+        );
+        Book book = bookRepository.findById(requestDto.bookId()).orElseThrow(
+                () -> new EntityNotFoundException("Cannot find book")
+        );
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItem.setQuantity(requestDto.quantity());
+        shoppingCart.getCartItems().add(cartItem);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto update(Long userId,
+                                  Long cartItemId,
+                                  CartItemUpdateRequestDto quantityDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find shopping cart by id " + userId)
+                );
+        CartItem cartItem = cartItemRepository
+                .findCartItemByIdAndShoppingCartId(cartItemId, shoppingCart.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find cart item by id " + cartItemId + "for user by id " + userId)
+                );
+        cartItem.setQuantity(quantityDto.quantity());
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(cartItem.getShoppingCart());
```

Comment:
With Transactional you may not call save method in repository. Check it

```suggestion
        cartItem.setQuantity(quantityDto.quantity());
        return shoppingCartMapper.toDto(shoppingCart);
```